substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -981,43 +995,88 @@ impl<T: Config> Pallet<T> {
 	///
 	/// If the nominator already exists, their nominations will be updated.
 	///
-	/// NOTE: you must ALWAYS use this function to add nominator or update their targets. Any access
-	/// to `Nominators` or `VoterList` outside of this function is almost certainly
-	/// wrong.
+	/// NOTE: you must ALWAYS use this function to add nominator or update their targets. Any
+	/// access to `Nominators` or `VoterList` outside of this function is almost certainly wrong.
 	pub fn do_add_nominator(who: &T::AccountId, nominations: Nominations<T>) {
-		if !Nominators::<T>::contains_key(who) {
-			// maybe update sorted list.
-			let _ = T::VoterList::on_insert(who.clone(), Self::weight_of(who))
-				.defensive_unwrap_or_default();
-		}
-		Nominators::<T>::insert(who, nominations);
+		match (Nominators::<T>::contains_key(who), T::VoterList::contains(who)) {
+			(false, false) => {
+				// new nomination
+				Nominators::<T>::insert(who, nominations);
+				T::EventListeners::on_nominator_add(who);
+			},
+			(true, true) | (false, true) => {
+				// update nominations or un-chill nominator.
+				let prev_nominations = Self::nominations(who).unwrap_or_default();
+				Nominators::<T>::insert(who, nominations);
+				T::EventListeners::on_nominator_update(who, prev_nominations);
+			},
+			(true, false) => {
+				defensive!("unexpected state.");
```

Comment:
Should we still add the new nominations to stake tracker?

Reply:
This arm should never be called, if everything is OK. a validator is always a nominator.